sel/term.hpp

```diff
@@ -1,11 +1,26 @@
 #pragma once
 
+#include "sel/strongly_ordered.hpp"
+
 #include <concepts>
 
 namespace sel {
 
+/// hook for types to declare usability as a term
+template <class T>
+constexpr bool enable_as_term = false;
+
 /// a symbol that represents a quantity in an expression
+///
+/// A type that models `term` must be `copyable` and must explicitly declare
+/// itself as usable as a `term` by specializing `enable_as_term`.
+///
+/// Additionally, all valid values of the type must be totally ordered in order
+/// to allow reordering of terms in an expression.
 template <class T>
-concept term = std::same_as<T, class variable>;
+concept term =               //
+    std::copyable<T> and     //
+    strongly_ordered<T> and  //
+    enable_as_term<T>;
```

Comment:
what is the motivation for the opt-in strategy, rather than accepting all copy and strongly ordered T?

Reply:
I'm not sure how many value types we want to support as `constant`s. `std::string` is copyable and strongly ordered, but I'm not sure we want that.

Reply:
what about foreign types, if someone uses a bigint lib or something?

Reply:
> what about foreign types, if someone uses a bigint lib and wants to use that?

I'm not sure if we should support that - but maybe it's worth doing in a separate PR. Do you want to try and get `boost::multiprecision` to work here?

https://www.boost.org/doc/libs/1_71_0/libs/math/doc/html/math_toolkit/high_precision/use_multiprecision.html

> Can they impl enable_as_term for types coming from another library?

Yes there's no orphan rule in C++ :).

Reply:
You can pull in boost from here:
https://registry.bazel.build/modules/boost.multiprecision
or here if necessary
https://github.com/nelhage/rules_boost

Probably from BCR will be easiest.

Here's an example where I've used it before:
https://github.com/oliverlee/rigid-geometric-algebra/blob/b80222d1d1b55a0552ee3c5dd2b4790addfcc77a/third_party/symengine.BUILD.bazel#L346